pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+		return err
+	}
+	s.setMidRowForPrevKey(key.Key)
+	err := s.maybeDoSizeFlushWithNextKey(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	if len(s.flushedFiles) == 0 {
+		panic(
+			"unexpected missing BackupManifest_File after flushes when writing key",
+		)
+	}
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if lastFile.EntryCounts.DataSize >= fileSpanByteLimit {
+		// As we are splitting due to the fileSpanByteLimit, the new span will need
+		// to contiguously extend the last span. We can use the current key as an
+		// exclusive end key for the last file and an inclusive start key for the
+		// new file.
+		newSpan := roachpb.Span{
+			Key:    key.Key,
+			EndKey: lastFile.Span.EndKey,
+		}
+		lastFile.Span.EndKey = newSpan.Key
+		if err := s.Reset(ctx, newSpan); err != nil {
+			return err
+		}
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	if err := s.maybeFlushOOOSpans(ctx, newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	var isContiguous, isSame bool
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		isSame = lastFile.Span.Equal(newSpan)
+		isContiguous = isContiguousSpan(lastFile.Span, newSpan)
+		if isSame {
+			return nil
+		} else if isContiguous &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	// If we are resetting to a new span that is not contiguous with the
+	// last span, we need to ensure that we are not mid-row.
+	if lastFile != nil && !isContiguous && s.midRow {
```

Comment:
i'm not sure why s.midRow is being evaluated in this complex conditional.  Could we just add a check at the top of Reset which is:
```
 if s.MidRow {
   return errors.AssertionaFailedf("cannot reset the sink if the last file is mid row")
 }
```

Reply:
>  s.midRow is being evaluated in this complex conditional

Answered in the prior comment

Reply:
Also just noticed that I checked `s.midRow` twice here. This is what's known as a double truth in the professional industry, very rarely seen in production code but it has its uses I'm sure.